### PR TITLE
addNewDeck() should throw exception if deckName already exists

### PR DESCRIPTION
**Pull request by marcardar**
Also fixes problem where dyn was being set in DECKS insert() and desc was being set for deck name that already exists.

### CODE REVIEW

**Comment by marcardar** [comment]
Just changed it so that we don't bother with the Deck lookup if the desc has not been set

**Comment by marcardar** [comment]
Actually, maybe you prefer to pass the desc into a new Decks.id(name, desc, create, type) method replacing Decks.id(name, create, type)?

**Comment by marcardar** [comment]
I made that change regarding Decks.id() accepting a desc - seems much better.

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java [hunk 1]
```diff
@@ -932,22 +932,13 @@ public Uri insert(Uri uri, ContentValues values) {
...
-                        }
-                    } catch (JSONException e) {
-                        Timber.e(e, "Could not set a field of new deck %s", deckName);
-                        return null;
-                    }
+                did = col.getDecks().id(deckName, false);
+                if (did != null) {
+                    throw new IllegalArgumentException("Deck name already exists: " + deckName);
+                }
+                did = col.getDecks().id(deckName, values.getAsString(FlashCardsContract.Deck.DECK_DESC));
```
Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: Does it make sense to put something here like:

```
Integer deckDyn = values.getAsInteger(FlashCardsContract.Deck.DECK_DYN);
if (deckDyn == 1) {
    Deck.newDyn(deckName);
}
```

Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: I don't think the second String argument of getDecks().id() is supposed to be for the deck description... I think you should use the old code for that.

Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: Ah I didn't see the changes to libanki below. I'd prefer not to change libanki and just set the JSON for that one.

Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: > Does it make sense to put something here like:

> ```
Integer deckDyn = values.getAsInteger(FlashCardsContract.Deck.DECK_DYN);
if (deckDyn == 1) {
    did = Decks.newDyn(deckName);
}
```

I don't think it's useful to allow this functionality at present as filtered decks require extra configuration parameters to be of any use. In fact I'm thinking that maybe we should skip filtered decks entirely from the deck list.

Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: It might be easier for me to make a PR actually

Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: > I'm thinking that maybe we should skip filtered decks entirely from the deck list.

Actually, this isn't a good idea. I had forgotten that decks are returned with their schedule information and used by AnkiWear which wants a full deck list. I guess we'll have to go through and protect all the forbidden operations with filtered decks then.

**Comment by timrae** [comment]
See #4177

**Comment by marcardar** [comment]
> I don't think the second String argument of getDecks().id() is supposed to be for the deck description... I think you should use the old code for that.

Yeah, well I checked where id(String, String) was used and it was only used in Decks.newDyn() so I changed that to use id(String, boolean, String). Generally, it would be much clearer to rename some of those methods (the ones that set create to true) from id() to newDeck().

**Comment by timrae** [comment]
> Generally, it would be much clearer to rename some of those methods

In principle we don't rename or add extra arguments to any methods in libanki as we try to keep it as consistent with the original python as possible.

**Comment by marcardar** [comment]
> In principle we don't rename or add extra arguments to any methods in libanki as we try to keep it as consistent with the original python as possible.

Noted, thanks.

**Pull request closed without merging**